File: Day12/main.js
function calculatePrice(ornaments) {
  const prices = {
    '*': 1, 'o': 5, '^': 10, '#': 50, '@': 100,
  }
  const total = Array.from(ornaments).reduceRight((sum, ornament, i)=>{
    const price = prices[ornament]
    const rightPrice = prices[ornaments[i+1]]
    return [sum + price, sum - price][(+(rightPrice > price))]
  }, 0)
  return [undefined, total][+!!total]
}

File: Day12/main.py
def calculate_price(ornaments):
    prices = {'*': 1, 'o': 5, '^': 10, '#': 50, '@': 100}
    total = 0

    for i in range(len(ornaments)):
        ornament = ornaments[i]
        if ornament not in prices:
            return None

        next_price = prices.get(ornaments[i + 1], 0) if i + 1 < len(ornaments) else 0
        total += prices[ornament] if prices[ornament] >= next_price else -prices[ornament]

    return total
